### ISSUE TITLE
Swallow function not working with ranger

### ISSUE THREAD

**Issue by og900aero** [ranger/ranger#2837]
- Operating system and version: Debian 11 (stable) 
- Terminal emulator and version: st 0.9
- Python version: 3.9.2
- Ranger version/commit: 1.9.3
- Locale:  en_US.UTF-8

I use DWM tilling window manager with swallow patch. This working well with lf, vifm, mc file managers when opening videos, pictures, but when I use ranger and open pictures, videos, then opens in a separate window (so swallow doesnt work).
Is there a setting that blocks the window manager's swallow function?

**Comment by toonn** [comment]
[This thread on our subreddit should help you out](https://www.reddit.com/r/ranger/comments/12jnxjc/window_swallowing_in_ranger/). It's a great place for questions that other users might be able to answer.

I do wonder why two users experienced this issue so close together, is it a new feature of the WM? Ranger doesn't apply any magic sauce depending on window managers so I'm fairly certain it's not caused by a change on our end.

**Comment by og900aero** [comment]
> [This thread on our subreddit should help you out](https://www.reddit.com/r/ranger/comments/12jnxjc/window_swallowing_in_ranger/). It's a great place for questions that other users might be able to answer.
> 
> I do wonder why two users experienced this issue so close together, is it a new feature of the WM? Ranger doesn't apply any magic sauce depending on window managers so I'm fairly certain it's not caused by a change on our end.

Thx your reply. I looked at the link and I was right, the ranger overrides the window manager swallow functions, so it definitely opens a new window, ergo ranger default settings is not good. The solution is:

edit rifle.conf and delete ", flag f" the following rows (is use mpv, and imv):

mime ^image, has imv-x11, X, flag f = imv-x11 -n "$1" "$PWD"
mime ^video,       has mpv,      X, flag f = mpv -- "$@"
mime ^video,       has mpv,      X, flag f = mpv --fs -- "$@"

**Comment by toonn** [comment]
No, Ranger does no such thing. I'm the person who suggested that solution. GUI applications like MPV don't make use of the terminal window so it is inconvenient for Ranger to be inoperative while they are running. The default behavior is to fork such applications so Ranger remains usable while they are running.

Your use case is niche and easily accomodated by changing the Rifle configuration file so we will not be changing the defaults.

I think ideally the WM would implement a feature to do window swallowing even if a process is forked, rather than exec'ed.

**Comment by og900aero** [comment]
> No, Ranger does no such thing. I'm the person who suggested that solution. GUI applications like MPV don't make use of the terminal window so it is inconvenient for Ranger to be inoperative while they are running. The default behavior is to fork such applications so Ranger remains usable while they are running.
> 
> Your use case is niche and easily accomodated by changing the Rifle configuration file so we will not be changing the defaults.
> 
> I think ideally the WM would implement a feature to do window swallowing even if a process is forked, rather than exec'ed.

I use a DWM that has a swallow function with patch, but I also wrote this because swallow works perfectly in all other file managers (vifm, joshuto, lf, nnn), except for ranger (except after deleting the described part). It wouldn't be bad if there was a warning about this in the ranger documentation and the solution.

**Comment by toonn** [comment]
I'm not sure Ranger should be responsible for documenting something like this. We don't really spell out differences in behavior for macOS or WSL users either. I think the WM should clearly document that swallowing doesn't work for forked processes and it's already clearly documented that `flag f` makes Rifle fork processes.
